runtime(i3config): add `all` option for popup_during_fullscreen

bruhtus opened the pull request:
It seems like since i3 version 4.24, popup_during_fullscreen has new option `all`.

So add `all` option for popup_during_fullscreen to prevent `all` option highlighted as error.

Reference:
https://i3wm.org/docs/userguide.html#_popups_during_fullscreen_mode

bruhtus committed:
runtime(i3config): add `all` option for popup_during_fullscreen

It seems like since i3 version 4.24, popup_during_fullscreen has new
option `all`.

So add `all` option for popup_during_fullscreen to prevent `all` option
highlighted as error.

Reference:
https://i3wm.org/docs/userguide.html#_popups_during_fullscreen_mode

Signed-off-by: Robertus Chris <robertusdchris@gmail.com>

chrisbra commented:
ping @litoj @jamespeapen

jamespeapen commented:
Also pinging @hiqua.

Just tested this. This works for i3, but would necessitate adding something like this to the sway syntax file so `all` is still highlighted as error; there may be a better way to restrict `all` to just i3.

```vim
" Present in i3 but not sway
syn keyword i3ConfigPopupFullscreenOpts smart ignore leave_fullscreen contained
```

It's unlikely that sway will add the `all` option: https://github.com/swaywm/sway/issues/8746.

bruhtus commented:
> It's unlikely that sway will add the all option: https://github.com/swaywm/sway/issues/8746.

Oh i thought that sway also implement the `all` option for popup_during_fullscreen, sorry about that 😅

Let me try if we can restrict the `all` option only in i3 config.

bruhtus committed:
runtime: add popup_during_fullscreen all in i3 but not in sway config

Seems like sway won't implement `all` option for
popup_during_fullscreen, so we need to make the `all` option for
popup_during_fullscreen in sway config highlight as an error.

Reference about sway not implement `all` option for
popup_during_fullscreen:
https://github.com/swaywm/sway/issues/8746

With that in mind, we can use syntax cluster in i3config.vim so that
we can combine the shared and extra options, and then when sourcing
i3config.vim in swayconfig.vim, we can remove the extra options from
the syntax cluster.

Signed-off-by: Robertus Chris <robertusdchris@gmail.com>

bruhtus commented:
I tried to split the options keyword into `i3ConfigPopupFullscreenShared` and `i3ConfigPopupFullscreenExtra`, and then combine those keyword in syntax cluster. When we are sourcing the i3 config with `runtime! syntax/i3config.vim` in `swayconfig.vim`, we can remove the `i3ConfigPopupFullscreenExtra` from the syntax cluster.

`i3ConfigPopupFullscreenShared` is for the popup_during_fullscreen options shared between i3 and sway, and `i3ConfigPopupFullscreenExtra` is for the popup_during_fullscreen options only in i3. If anyone has a better naming scheme, please let me know.

What do you think? 🤔

bruhtus committed:
chore(i3config/swayconfig): update link to new repo

Also increase the version of swayconfig to be the same as the original
plugin.

Signed-off-by: Robertus Chris <robertusdchris@gmail.com>

bruhtus committed:
runtime: add i3 opts extra in highlight default link

Also use the same name for syntax keyword and highlight default link
name.

Signed-off-by: Robertus Chris <robertusdchris@gmail.com>

bruhtus commented:
Should we link the highlight for `i3ConfigPopupFullscreenOptsExtra` with `i3ConfigPopupFullscreenOpts` instead of `i3ConfigOption`?

So that the user only need to change the highlight `i3ConfigPopupFullscreenOpts` and the extra opts will be changed too 🤔

bruhtus committed:
runtime(i3config): link popup fullscreen extra opts with opts

Signed-off-by: Robertus Chris <robertusdchris@gmail.com>

litoj commented:
Yes, looks and works all well.

chrisbra commented:
thanks all
